### ISSUE TITLE
handleSetDifferentColors fails for subsequent calls

### ISSUE THREAD

**Issue by dale3h** [toblum/McLighting#332]
# Problem/Motivation

I was playing around with the WebSocket protocol and could not figure out why multi LED and range LED commands were not working properly. The first call to `+[numled][hexrgb] [...]` succeeds, but any call thereafter will use the R value for G.

## Helpful information

| Helpful | Information |
| --- | --- |
| **Physical Board** | Wemos D1 Mini (4MB) |
| **OS** | macOS Mojave 10.14 |
| **Arduino IDE** | 1.8.8 |
| **Board Repo** | esp8266 by ESP8266 Community version 2.5.0 |
| **McLighting Version** | 2.2.0 |
| **WebSocket Client** | Smart Websocket Client (Chrome app) |

## Expected behavior

All calls to WebSocket protocols for multi LED and range LED to use proper R, G, and B values.

## Actual behavior

- All calls *excluding first call* to multi LED command use R value in place of G value. i.e. color `aabbcc` will be interpreted as `aaaacc`
- All calls *including first call* to range LED command use R value in place of G value. i.e. color `aabbcc` will be interpreted as `aaaacc`

### Payload: `+10aabbcc+11ddeeff`
```
WS: [0] Connected from 192.168.1.50 url: /
WS: [0] get Text: +10aabbcc+11ddeeff
led value: [10]. Entry threshold: <= [36] (=> 10aabbcc)
rgb.red: [aaP10] rgb.green: [bb⸮aaP10] rgb.blue: [cc⸮bb⸮aaP10]
rgb.red: [170] rgb.green: [187] rgb.blue: [204]
WS: Set single led [10] to [170] [187] [204] (10aabbcc)!
led value: [11]. Entry threshold: <= [36] (=> 11ddeeff)
rgb.red: [ddP11] rgb.green: [ee⸮ddP11] rgb.blue: [ff⸮ee⸮ddP11]
rgb.red: [221] rgb.green: [238] rgb.blue: [255]
WS: Set single led [11] to [221] [238] [255] (11ddeeff)!
WS: Set multiple LEDs in given color [+10aabbcc]
WS: [0] get Text: +10aabbcc+11ddeeff
led value: [10]. Entry threshold: <= [36] (=> 10aabbcc)
rgb.red: [aa10] rgb.green: [bb0aa10] rgb.blue: [cc⸮bb0aa10]
rgb.red: [170] rgb.green: [766122] rgb.blue: [204]
WS: Set single led [10] to [170] [170] [204] (10aabbcc)!
led value: [11]. Entry threshold: <= [36] (=> 11ddeeff)
rgb.red: [dd11] rgb.green: [ee0dd11] rgb.blue: [ff⸮ee0dd11]
rgb.red: [221] rgb.green: [975069] rgb.blue: [255]
WS: Set single led [11] to [221] [221] [255] (11ddeeff)!
WS: Set multiple LEDs in given color [+10aabbcc]
```

### Payload: `R0009aabbccR1019ddeeff`
```
WS: [0] Connected from 192.168.1.50 url: /
WS: [0] get Text: R0009aabbccR1019ddeeff
Setting RANGE from [0] to [9] as color [aabbcc] 
led value: [0]. Entry threshold: <= [36] (=> 00aabbcc)
rgb.red: [aaP00] rgb.green: [bb0aaP00] rgb.blue: [cc]
rgb.red: [170] rgb.green: [766122] rgb.blue: [204]
WS: Set single led [0] to [170] [170] [204] (00aabbcc)!
led value: [1]. Entry threshold: <= [36] (=> 01aabbcc)
rgb.red: [aaP01] rgb.green: [bb0aaP01] rgb.blue: [cc]
rgb.red: [170] rgb.green: [766122] rgb.blue: [204]
WS: Set single led [1] to [170] [170] [204] (01aabbcc)!
led value: [2]. Entry threshold: <= [36] (=> 02aabbcc)
rgb.red: [aaP02] rgb.green: [bb0aaP02] rgb.blue: [cc]
rgb.red: [170] rgb.green: [766122] rgb.blue: [204]
WS: Set single led [2] to [170] [170] [204] (02aabbcc)!
led value: [3]. Entry threshold: <= [36] (=> 03aabbcc)
rgb.red: [aaP03] rgb.green: [bb0aaP03] rgb.blue: [cc]
rgb.red: [170] rgb.green: [766122] rgb.blue: [204]
WS: Set single led [3] to [170] [170] [204] (03aabbcc)!
led value: [4]. Entry threshold: <= [36] (=> 04aabbcc)
rgb.red: [aaP04] rgb.green: [bb0aaP04] rgb.blue: [cc]
rgb.red: [170] rgb.green: [766122] rgb.blue: [204]
WS: Set single led [4] to [170] [170] [204] (04aabbcc)!
led value: [5]. Entry threshold: <= [36] (=> 05aabbcc)
rgb.red: [aaP05] rgb.green: [bb0aaP05] rgb.blue: [cc]
rgb.red: [170] rgb.green: [766122] rgb.blue: [204]
WS: Set single led [5] to [170] [170] [204] (05aabbcc)!
led value: [6]. Entry threshold: <= [36] (=> 06aabbcc)
rgb.red: [aaP06] rgb.green: [bb0aaP06] rgb.blue: [cc]
rgb.red: [170] rgb.green: [766122] rgb.blue: [204]
WS: Set single led [6] to [170] [170] [204] (06aabbcc)!
led value: [7]. Entry threshold: <= [36] (=> 07aabbcc)
rgb.red: [aaP07] rgb.green: [bb0aaP07] rgb.blue: [cc]
rgb.red: [170] rgb.green: [766122] rgb.blue: [204]
WS: Set single led [7] to [170] [170] [204] (07aabbcc)!
led value: [8]. Entry threshold: <= [36] (=> 08aabbcc)
rgb.red: [aaP08] rgb.green: [bb0aaP08] rgb.blue: [cc]
rgb.red: [170] rgb.green: [766122] rgb.blue: [204]
WS: Set single led [8] to [170] [170] [204] (08aabbcc)!
led value: [9]. Entry threshold: <= [36] (=> 09aabbcc)
rgb.red: [aaP09] rgb.green: [bb0aaP09] rgb.blue: [cc]
rgb.red: [170] rgb.green: [766122] rgb.blue: [204]
WS: Set single led [9] to [170] [170] [204] (09aabbcc)!
Setting RANGE from [10] to [19] as color [ddeeff] 
led value: [10]. Entry threshold: <= [36] (=> 10ddeeff)
rgb.red: [ddP10] rgb.green: [ee0ddP10] rgb.blue: [ff]
rgb.red: [221] rgb.green: [975069] rgb.blue: [255]
WS: Set single led [10] to [221] [221] [255] (10ddeeff)!
led value: [11]. Entry threshold: <= [36] (=> 11ddeeff)
rgb.red: [ddP11] rgb.green: [ee0ddP11] rgb.blue: [ff]
rgb.red: [221] rgb.green: [975069] rgb.blue: [255]
WS: Set single led [11] to [221] [221] [255] (11ddeeff)!
led value: [12]. Entry threshold: <= [36] (=> 12ddeeff)
rgb.red: [ddP12] rgb.green: [ee0ddP12] rgb.blue: [ff]
rgb.red: [221] rgb.green: [975069] rgb.blue: [255]
WS: Set single led [12] to [221] [221] [255] (12ddeeff)!
led value: [13]. Entry threshold: <= [36] (=> 13ddeeff)
rgb.red: [ddP13] rgb.green: [ee0ddP13] rgb.blue: [ff]
rgb.red: [221] rgb.green: [975069] rgb.blue: [255]
WS: Set single led [13] to [221] [221] [255] (13ddeeff)!
led value: [14]. Entry threshold: <= [36] (=> 14ddeeff)
rgb.red: [ddP14] rgb.green: [ee0ddP14] rgb.blue: [ff]
rgb.red: [221] rgb.green: [975069] rgb.blue: [255]
WS: Set single led [14] to [221] [221] [255] (14ddeeff)!
led value: [15]. Entry threshold: <= [36] (=> 15ddeeff)
rgb.red: [ddP15] rgb.green: [ee0ddP15] rgb.blue: [ff]
rgb.red: [221] rgb.green: [975069] rgb.blue: [255]
WS: Set single led [15] to [221] [221] [255] (15ddeeff)!
led value: [16]. Entry threshold: <= [36] (=> 16ddeeff)
rgb.red: [ddP16] rgb.green: [ee0ddP16] rgb.blue: [ff]
rgb.red: [221] rgb.green: [975069] rgb.blue: [255]
WS: Set single led [16] to [221] [221] [255] (16ddeeff)!
led value: [17]. Entry threshold: <= [36] (=> 17ddeeff)
rgb.red: [ddP17] rgb.green: [ee0ddP17] rgb.blue: [ff]
rgb.red: [221] rgb.green: [975069] rgb.blue: [255]
WS: Set single led [17] to [221] [221] [255] (17ddeeff)!
led value: [18]. Entry threshold: <= [36] (=> 18ddeeff)
rgb.red: [ddP18] rgb.green: [ee0ddP18] rgb.blue: [ff]
rgb.red: [221] rgb.green: [975069] rgb.blue: [255]
WS: Set single led [18] to [221] [221] [255] (18ddeeff)!
led value: [19]. Entry threshold: <= [36] (=> 19ddeeff)
rgb.red: [ddP19] rgb.green: [ee0ddP19] rgb.blue: [ff]
rgb.red: [221] rgb.green: [975069] rgb.blue: [255]
WS: Set single led [19] to [221] [221] [255] (19ddeeff)!
WS: Set range of LEDs in given color [R0009aabbcc]
```

The Arduino was soft-reset before sending each payload. A hard-reset produces the exact same results.

### Light multiple LEDs in the given colors

1. Connect to the McLighting WebSocket server. *The WebSocket server runs at `ws://ESP_IP_ADDRESS_OR_HOSTNAME:81/`*
1. Send payload `+10aabbcc+11ddeeff` (does behave as expected)
1. Send payload `+10aabbcc+11ddeeff` (does ***not*** behave as expected)

### Light multiple LED ranges in the given colors

1. Connect to the McLighting WebSocket server. *The WebSocket server runs at `ws://ESP_IP_ADDRESS_OR_HOSTNAME:81/`*
1. Send payload `R0009aabbccR1019ddeeff` (does ***not*** behave as expected)

## Proposed changes

I traced the code to where I think this might be happening, but nothing obvious stood out to me. I suspect that it may be a pointer corruption/overwrite issue.9

**Comment by dale3h** [comment]
Null-terminating the hex char arrays in `handleSetSingleLED()` fixes this issue:

```
    redhex[2] = 0x00;
    greenhex[2] = 0x00;
    bluehex[2] = 0x00;
```

I have submitted a PR (#333) to fix this issue.

**Comment by dale3h** [comment]
Closed via #333
